libs/@local/hash-graph-types/typescript/src/entity.ts

```diff
@@ -83,22 +83,56 @@ export type PropertyMetadataElement =
   | PropertyMetadataObject
   | PropertyMetadataValue;
 
+export const isValueMetadata = (
+  metadata: PropertyMetadataElement,
+): metadata is PropertyMetadataValue => !("value" in metadata);
+
+export const isArrayMetadata = (
+  metadata: PropertyMetadataElement,
+): metadata is PropertyMetadataArray =>
+  !isValueMetadata(metadata) && Array.isArray(metadata.value);
+
+export const isObjectMetadata = (
+  metadata: PropertyMetadataElement,
+): metadata is PropertyMetadataObject =>
+  !isValueMetadata(metadata) && !Array.isArray(metadata.value);
+
 export interface PropertyMetadataValue {
   metadata: Omit<ValueMetadata, "dataTypeId"> & {
     dataTypeId?: VersionedUrl;
   };
 }
 
 export interface PropertyMetadataObject {
-  value?: Record<BaseUrl, PropertyMetadataElement>;
+  value: Record<BaseUrl, PropertyMetadataElement>;
   metadata?: ObjectMetadata;
 }
 
 export interface PropertyMetadataArray {
-  value?: PropertyMetadataElement[];
+  value: PropertyMetadataElement[];
   metadata?: ArrayMetadata;
 }
 
+export type PropertyWithMetadata =
+  | PropertyArrayWithMetadata
+  | PropertyObjectWithMetadata
+  | PropertyValueWithMetadata;
+
+export interface PropertyValueWithMetadata {
+  value: JsonValue;
+  metadata: PropertyMetadataValue["metadata"];
+}
+
+export interface PropertyObjectWithMetadata {
+  value: Record<BaseUrl, PropertyWithMetadata>;
+  metadata?: PropertyMetadataObject["metadata"];
+}
+
+export interface PropertyArrayWithMetadata {
+  value: PropertyWithMetadata[];
+  metadata?: PropertyMetadataArray["metadata"];
+}
+
```

Comment:
@CiaranMn do you have any opinions on the type names? I think making them consistent everywhere would be good. We currently also have `EntityPropertiesObject` which is the same as `PropertyObjectWithMetadata` but without metadata, so maybe rename those to `PropertyObject`?

Reply:
As these are used quite a lot I'm going to change the naming in a follow-up to avoid conflicts. We can discuss the names either here or in a later PR. Changing them in the head branch is easy.

Reply:
I think fine to remove `Entity`

Reply:
I consistently renamed everything (Rust and Javascript - but I left BP untouched)